packages/vite/src/node/build.ts

```diff
@@ -639,6 +652,34 @@ export async function build(
         }
       })
 
+      // watch config related files in another chokidar instance to
+      // re-run the build function to start a new rollup watcher
+      if (config.build.watchConfig) {
+        const chokidar = await import('chokidar')
+        const configFileChokidar = chokidar.watch(
+          (
+            [
+              config.configFile,
+              ...config.configFileDependencies,
+              '.env',
+              '.env.*',
+            ].filter(Boolean) as string[]
+          ).map(resolve),
+          {
+            ignoreInitial: true,
+          },
+        )
+
+        configFileChokidar.on('all', async () => {
+          await Promise.all([watcher.close(), configFileChokidar.close()])
+          build(inlineConfig)
+        })
```

Comment:
Maybe it would be nice to log that we're restarting because of config change, similar to

https://github.com/vitejs/vite/blob/7110ddf39b3b6f122945bbbc5025773c64f9d948/packages/vite/src/node/server/hmr.ts#L59-L66